src/main/java/org/openrewrite/sql/FormatSql.java

```diff
@@ -0,0 +1,190 @@
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class FormatSql extends Recipe {
+    @Option(displayName = "SQL dialect to be used to format SQL snippets.", description = "", example = "PostgreSql", required = false)
+    @Nullable
+    String sqlDialect;
+
+    public FormatSql() {
+        sqlDialect = Dialect.StandardSql.name();
+    }
+
+    public FormatSql(String sqlDialect) {
+        this.sqlDialect = sqlDialect;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Format SQL in String Text Blocks";
+    }
+
+    @Override
+    public String getDescription() {
+        return "...";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new UsesJavaVersion<>(17);
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.Literal visitLiteral(J.Literal lit, ExecutionContext ctx) {
+                J.Literal literal = super.visitLiteral(lit, ctx);
+                final SqlDetector detector = new SqlDetector();
+
+                if (isTextBlock(literal)) {
+                    String value = literal.getValue().toString();
+
+                    if (detector.isSql(value)) {
+                        String formatted = SqlFormatter.of(Dialect.valueOf(sqlDialect)).format(value);
+
+                        TabsAndIndentsStyle tabsAndIndentsStyle = Optional
+                                .ofNullable(getCursor().firstEnclosingOrThrow(SourceFile.class)
+                                        .getStyle(TabsAndIndentsStyle.class))
+                                .orElse(IntelliJ.tabsAndIndents());
+                        boolean useTab = tabsAndIndentsStyle.getUseTabCharacter();
+                        int tabSize = tabsAndIndentsStyle.getTabSize();
+
+                        String indentation = getIndents(literal.getValueSource(), useTab, tabSize);
+
+                        // preserve trailing spaces
+                        formatted = formatted.replace(" \n", "\\s\n");
+                        // handle preceding indentation
+                        formatted = formatted.replace("\n", "\n" + indentation);
+
+                        // add first line
+                        formatted = "\n" + indentation + formatted;
+
+                        // add last line to ensure the closing delimiter is in a new line to manage
+                        // indentation & remove theformatted
+                        // need to escape ending quote in the content
+                        boolean isEndsWithNewLine = literal.getValueSource().endsWith("\n");
+                        if (!isEndsWithNewLine) {
+                            formatted = formatted + "\\\n" + indentation;
+                        }
+
+                        return literal.withValueSource(String.format("\"\"\"%s\"\"\"", formatted));
```

Comment:
This is basically what causes the test to fail. The recipe "modifies" the LST on a second cycle with a no-op modification. As checking for modifications using `Object#equals(Object)` would be too expensive, OpenRewrite uses referential equality. This in turn means that recipes should try hard to only modify the LST when necessary.

```suggestion
                        String newValueSource = String.format("\"\"\"%s\"\"\"", formatted);
                        if ( !newValueSource.equals(literal.getValueSource())) {
                            return literal.withValueSource(newValueSource);
                        }
```

Reply:
I'm not fully understanding your comment. Maybe the maintainers do and can commit the change.

Reply:
Applied in https://github.com/openrewrite/rewrite-sql/pull/2/commits/64b27fda1489affab52245f1bf023ab4e65a8f68